aiokafka/producer/message_accumulator.py

```diff
@@ -236,3 +247,47 @@ def drain_by_nodes(self, ignore_nodes):
         self._wait_data_future = create_future(loop=self._loop)
 
         return nodes, unknown_leaders_exist
+
+    def create_builder(self):
+        magic = 0 if self._api_version < (0, 10) else 1
+        return BatchBuilder(magic, self._batch_size, self._compression_type)
+
+    def _append_batch(self, builder, tp):
+        batch = MessageBatch(tp, builder, self._batch_ttl, self._loop)
+        self._batches[tp].append(batch)
+        if not self._wait_data_future.done():
+            self._wait_data_future.set_result(None)
+        return batch
+
+    @asyncio.coroutine
+    def add_batch(self, builder, tp, timeout):
+        """Add BatchBuilder to queue by topic-partition.
+
+        Arguments:
+            builder (BatchBuilder): batch object to enqueue.
+            tp (TopicPartition): topic and partition to enqueue this batch for.
+            timeout (int): time in seconds to wait for a free slot in the batch
+                queue.
+
+        Returns:
+            MessageBatch: delivery wrapper around the BatchBuilder object.
+                to the broker.
+
+        Raises:
+            aiokafka.errors.ProducerClosed: the accumulator has already been
+                closed and flushed.
+            aiokafka.errors.KafkaTimeoutError: the batch could not be added
+                within the specified timeout.
+        """
+        if self._closed:
+            raise ProducerClosed()
+
+        while timeout > 0:
+            start = self._loop.time()
+            pending = self._batches.get(tp)
+            if pending:
+                yield from pending[-1].wait_drain()
```

Comment:
Pass timeout to `wait_drain()`. See my snippet above, it already has the timeout passed.